// @migroscomtr/api.module.ts
import { NgModule, ModuleWithProviders, SkipSelf, Optional } from '@angular/core';
import { Configuration } from './configuration';
import { HttpClient } from '@angular/common/http';

import { AccountControllerService } from './api/accountController.service';
import { AccountControllerV2Service } from './api/accountControllerV2.service';
import { AddressRestControllerService } from './api/addressRestController.service';
import { AgreementRestControllerService } from './api/agreementRestController.service';
import { AnonymousCheckoutRestControllerService } from './api/anonymousCheckoutRestController.service';
import { BannerRestControllerService } from './api/bannerRestController.service';
import { BrandRestControllerService } from './api/brandRestController.service';
import { CampaignRestControllerService } from './api/campaignRestController.service';
import { CartRestControllerService } from './api/cartRestController.service';
import { CartScreenControllerService } from './api/cartScreenController.service';
import { CategoryRestControllerService } from './api/categoryRestController.service';
import { CategorySearchResultScreenControllerService } from './api/categorySearchResultScreenController.service';
import { CheckoutRestControllerService } from './api/checkoutRestController.service';
import { CheckoutScreenControllerService } from './api/checkoutScreenController.service';
import { CompanyGroupRestControllerService } from './api/companyGroupRestController.service';
import { CompatibilityRestControllerService } from './api/compatibilityRestController.service';
import { DeliveryRestControllerService } from './api/deliveryRestController.service';
import { DeviceRestControllerService } from './api/deviceRestController.service';
import { DynamicPageRestControllerService } from './api/dynamicPageRestController.service';
import { EmailUnsubscribeControllerService } from './api/emailUnsubscribeController.service';
import { ExternalCatalogRestControllerService } from './api/externalCatalogRestController.service';
import { ExternalCompanyRestControllerService } from './api/externalCompanyRestController.service';
import { ExternalWarehouseService } from './api/externalWarehouse.service';
import { FaqRestControllerService } from './api/faqRestController.service';
import { FavouriteProductRestControllerService } from './api/favouriteProductRestController.service';
import { FeedbackRestControllerService } from './api/feedbackRestController.service';
import { HomeScreenControllerService } from './api/homeScreenController.service';
import { KurbanRestControllerService } from './api/kurbanRestController.service';
import { LocationRestControllerService } from './api/locationRestController.service';
import { LoyaltyProgramsControllerService } from './api/loyaltyProgramsController.service';
import { MoneyPayRestControllerService } from './api/moneyPayRestController.service';
import { MoneyTransferRestControllerService } from './api/moneyTransferRestController.service';
import { OrderAnonymousRestControllerService } from './api/orderAnonymousRestController.service';
import { OrderRestControllerService } from './api/orderRestController.service';
import { OrderScreenControllerService } from './api/orderScreenController.service';
import { OtpAuthenticationControllerService } from './api/otpAuthenticationController.service';
import { OtpRegistrationControllerService } from './api/otpRegistrationController.service';
import { OtpRegistrationControllerV2Service } from './api/otpRegistrationControllerV2.service';
import { PaymentCallbackRestControllerService } from './api/paymentCallbackRestController.service';
import { PaymentRestControllerService } from './api/paymentRestController.service';
import { ProductRestControllerService } from './api/productRestController.service';
import { ProductScreenControllerService } from './api/productScreenController.service';
import { ProductSearchAutoCompleteRestControllerService } from './api/productSearchAutoCompleteRestController.service';
import { ProductSearchResultScreenControllerService } from './api/productSearchResultScreenController.service';
import { PushNotificationRestControllerService } from './api/pushNotificationRestController.service';
import { RecipeRestControllerService } from './api/recipeRestController.service';
import { RegionRestControllerService } from './api/regionRestController.service';
import { ScreenViewRestControllerService } from './api/screenViewRestController.service';
import { ServerErrorControllerService } from './api/serverErrorController.service';
import { SettingPackRestControllerService } from './api/settingPackRestController.service';
import { ShoppingListSearchResultScreenControllerService } from './api/shoppingListSearchResultScreenController.service';
import { ShoppingListSystemRestControllerService } from './api/shoppingListSystemRestController.service';
import { ShoppingListUserRestControllerService } from './api/shoppingListUserRestController.service';
import { StatepackControllerService } from './api/statepackController.service';
import { StatepackRestControllerService } from './api/statepackRestController.service';
import { TransitionAuthenticationControllerService } from './api/transitionAuthenticationController.service';
import { TrendKeywordControllerService } from './api/trendKeywordController.service';
import { UserRestControllerService } from './api/userRestController.service';
import { UtmRestControllerService } from './api/utmRestController.service';

@NgModule({
  imports:      [],
  declarations: [],
  exports:      [],
  providers: []
})
export class ApiModule {
    public static forRoot(configurationFactory: () => Configuration): ModuleWithProviders<ApiModule> {
        return {
            ngModule: ApiModule,
            providers: [ { provide: Configuration, useFactory: configurationFactory } ]
        };
    }

    constructor( @Optional() @SkipSelf() parentModule: ApiModule,
                 @Optional() http: HttpClient) {
        if (parentModule) {
            throw new Error('ApiModule is already loaded. Import in your base AppModule only.');
        }
        if (!http) {
            throw new Error('You need to import the HttpClientModule in your AppModule! \n' +
            'See also https://github.com/angular/angular/issues/20575');
        }
    }
}
