// @migroscomtr/model/acceptAgreementRequest.model.ts
/**
 * Api Documentation
 * Api Documentation
 *
 * The version of the OpenAPI document: 1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export interface AcceptAgreementRequest { 
    acceptedAgreements?: Array<AcceptAgreementRequest.AcceptedAgreementsEnum>;
    email?: string;
    token?: string;
}
export namespace AcceptAgreementRequest {
    export type AcceptedAgreementsEnum = 'MEMBERSHIP_AGREEMENT' | 'MEMBERSHIP_V2_AGREEMENT' | 'EXPLICIT_CONSENT_AGREEMENT' | 'DISTANT_SALES_AGREEMENT' | 'DISTANT_SALES_AGREEMENT_FORM' | 'PRIVACY_AGREEMENT' | 'PERSONAL_DATA_AGREEMENT' | 'MCC_MEMBERSHIP_AGREEMENT' | 'MONEY_KVK_MEMBERSHIP_INFO_AGREEMENT' | 'MONEY_KVK_MEMBERSHIP_AGREEMENT' | 'KVK_MEMBERSHIP_AGREEMENT' | 'MCC_TERM_OF_USE_AGREEMENT' | 'MONEY_GOLD_MEMBERSHIP_INFO_AGREEMENT' | 'MONEY_GOLD_MEMBERSHIP_AGREEMENT' | 'MONEY_GOLD_TERM_OF_USE_AGREEMENT';
    export const AcceptedAgreementsEnum = {
        MembershipAgreement: 'MEMBERSHIP_AGREEMENT' as AcceptedAgreementsEnum,
        MembershipV2Agreement: 'MEMBERSHIP_V2_AGREEMENT' as AcceptedAgreementsEnum,
        ExplicitConsentAgreement: 'EXPLICIT_CONSENT_AGREEMENT' as AcceptedAgreementsEnum,
        DistantSalesAgreement: 'DISTANT_SALES_AGREEMENT' as AcceptedAgreementsEnum,
        DistantSalesAgreementForm: 'DISTANT_SALES_AGREEMENT_FORM' as AcceptedAgreementsEnum,
        PrivacyAgreement: 'PRIVACY_AGREEMENT' as AcceptedAgreementsEnum,
        PersonalDataAgreement: 'PERSONAL_DATA_AGREEMENT' as AcceptedAgreementsEnum,
        MccMembershipAgreement: 'MCC_MEMBERSHIP_AGREEMENT' as AcceptedAgreementsEnum,
        MoneyKvkMembershipInfoAgreement: 'MONEY_KVK_MEMBERSHIP_INFO_AGREEMENT' as AcceptedAgreementsEnum,
        MoneyKvkMembershipAgreement: 'MONEY_KVK_MEMBERSHIP_AGREEMENT' as AcceptedAgreementsEnum,
        KvkMembershipAgreement: 'KVK_MEMBERSHIP_AGREEMENT' as AcceptedAgreementsEnum,
        MccTermOfUseAgreement: 'MCC_TERM_OF_USE_AGREEMENT' as AcceptedAgreementsEnum,
        MoneyGoldMembershipInfoAgreement: 'MONEY_GOLD_MEMBERSHIP_INFO_AGREEMENT' as AcceptedAgreementsEnum,
        MoneyGoldMembershipAgreement: 'MONEY_GOLD_MEMBERSHIP_AGREEMENT' as AcceptedAgreementsEnum,
        MoneyGoldTermOfUseAgreement: 'MONEY_GOLD_TERM_OF_USE_AGREEMENT' as AcceptedAgreementsEnum
    };
}
