// @migroscomtr/model/checkoutPaymentFormBean.model.ts
/**
 * Api Documentation
 * Api Documentation
 *
 * The version of the OpenAPI document: 1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export interface CheckoutPaymentFormBean { 
    alternativeProductChoice?: CheckoutPaymentFormBean.AlternativeProductChoiceEnum;
    bagSelected?: boolean;
    balanceName?: string;
    balanceType?: CheckoutPaymentFormBean.BalanceTypeEnum;
    balanceValue?: string;
    contactless: boolean;
    note?: string;
    paymentType: CheckoutPaymentFormBean.PaymentTypeEnum;
    ringDoorbell: boolean;
}
export namespace CheckoutPaymentFormBean {
    export type AlternativeProductChoiceEnum = 'NO_ALTERNATIVE' | 'UP_TO_CUSTOMER' | 'UP_TO_PERSONNEL';
    export const AlternativeProductChoiceEnum = {
        NoAlternative: 'NO_ALTERNATIVE' as AlternativeProductChoiceEnum,
        UpToCustomer: 'UP_TO_CUSTOMER' as AlternativeProductChoiceEnum,
        UpToPersonnel: 'UP_TO_PERSONNEL' as AlternativeProductChoiceEnum
    };
    export type BalanceTypeEnum = 'BKM' | 'CARD_REWARD' | 'CASH_ON_DELIVERY' | 'COUPON' | 'CREDIT_CARD' | 'CREDIT_CARD_ON_DELIVERY' | 'CUSTOMER_BOND' | 'GARANTI_PAY' | 'INSTANT_DISCOUNT' | 'LOAN' | 'MASTERPASS' | 'MONEY_POINT' | 'PERSONNEL_BOND' | 'VALUE_DATE' | 'VOUCHER' | 'WALLET';
    export const BalanceTypeEnum = {
        Bkm: 'BKM' as BalanceTypeEnum,
        CardReward: 'CARD_REWARD' as BalanceTypeEnum,
        CashOnDelivery: 'CASH_ON_DELIVERY' as BalanceTypeEnum,
        Coupon: 'COUPON' as BalanceTypeEnum,
        CreditCard: 'CREDIT_CARD' as BalanceTypeEnum,
        CreditCardOnDelivery: 'CREDIT_CARD_ON_DELIVERY' as BalanceTypeEnum,
        CustomerBond: 'CUSTOMER_BOND' as BalanceTypeEnum,
        GarantiPay: 'GARANTI_PAY' as BalanceTypeEnum,
        InstantDiscount: 'INSTANT_DISCOUNT' as BalanceTypeEnum,
        Loan: 'LOAN' as BalanceTypeEnum,
        Masterpass: 'MASTERPASS' as BalanceTypeEnum,
        MoneyPoint: 'MONEY_POINT' as BalanceTypeEnum,
        PersonnelBond: 'PERSONNEL_BOND' as BalanceTypeEnum,
        ValueDate: 'VALUE_DATE' as BalanceTypeEnum,
        Voucher: 'VOUCHER' as BalanceTypeEnum,
        Wallet: 'WALLET' as BalanceTypeEnum
    };
    export type PaymentTypeEnum = 'CASH_ON_DELIVERY' | 'CREDIT_CARD_ON_DELIVERY' | 'CREDIT_CARD' | 'MASTERPASS' | 'GARANTI_PAY' | 'BKM' | 'VALUE_DATE' | 'WALLET' | 'LOAN' | 'MONEY_POINT' | 'PERSONNEL_BOND' | 'CUSTOMER_BOND' | 'VOUCHER' | 'CARD_REWARD' | 'COUPON' | 'INSTANT_DISCOUNT';
    export const PaymentTypeEnum = {
        CashOnDelivery: 'CASH_ON_DELIVERY' as PaymentTypeEnum,
        CreditCardOnDelivery: 'CREDIT_CARD_ON_DELIVERY' as PaymentTypeEnum,
        CreditCard: 'CREDIT_CARD' as PaymentTypeEnum,
        Masterpass: 'MASTERPASS' as PaymentTypeEnum,
        GarantiPay: 'GARANTI_PAY' as PaymentTypeEnum,
        Bkm: 'BKM' as PaymentTypeEnum,
        ValueDate: 'VALUE_DATE' as PaymentTypeEnum,
        Wallet: 'WALLET' as PaymentTypeEnum,
        Loan: 'LOAN' as PaymentTypeEnum,
        MoneyPoint: 'MONEY_POINT' as PaymentTypeEnum,
        PersonnelBond: 'PERSONNEL_BOND' as PaymentTypeEnum,
        CustomerBond: 'CUSTOMER_BOND' as PaymentTypeEnum,
        Voucher: 'VOUCHER' as PaymentTypeEnum,
        CardReward: 'CARD_REWARD' as PaymentTypeEnum,
        Coupon: 'COUPON' as PaymentTypeEnum,
        InstantDiscount: 'INSTANT_DISCOUNT' as PaymentTypeEnum
    };
}
