// @migroscomtr/model/orderDTO.model.ts
/**
 * Api Documentation
 * Api Documentation
 *
 * The version of the OpenAPI document: 1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SimpleOrderFeedbackDTO } from './simpleOrderFeedbackDTO.model';
import { EtaInfo } from './etaInfo.model';
import { PaymentOption } from './paymentOption.model';


export interface OrderDTO { 
    active?: boolean;
    additionalOrderIds?: string;
    alternativeProductChoice?: OrderDTO.AlternativeProductChoiceEnum;
    alternativeRegionId?: number;
    alternativeStoreId?: number;
    anonymous?: boolean;
    bagCount?: number;
    bagPrice?: number;
    bagSelected?: boolean;
    cancelReason?: string;
    cancelled?: boolean;
    cancelledAt?: string;
    cancellerId?: number;
    channel?: OrderDTO.ChannelEnum;
    closedPaymentTypes?: string;
    collectionStatus?: OrderDTO.CollectionStatusEnum;
    companyId?: number;
    completed?: boolean;
    createdAt?: string;
    creationDate?: string;
    creatorId?: number;
    crmTransactionId?: string;
    deliveredAt?: string;
    deliveredPrice?: number;
    delivererRate?: number;
    delivererStoreId?: number;
    deliveryAddressDetails?: string;
    deliveryAddressId?: number;
    deliveryAddressObjectType?: string;
    deliveryAddressPhone?: string;
    deliveryDate?: string;
    deliveryId?: number;
    deliveryIdV2?: number;
    deliveryModel?: OrderDTO.DeliveryModelEnum;
    deliveryPrice?: number;
    deliveryStatus?: string;
    deliveryStatusV2?: string;
    deviceChannel?: string;
    deviceIdentifier?: string;
    devicePlatform?: string;
    deviceType?: string;
    deviceVersion?: string;
    disabledPaymentTypes?: string;
    disabledPayments?: string;
    districtId?: number;
    email?: string;
    embeddedCollectionInfo?: string;
    embeddedDeliveryAddressInfo?: string;
    embeddedEtaInfo?: string;
    embeddedGeolocationInfo?: string;
    embeddedInvoiceAddressInfo?: string;
    embeddedMinCartAmountDiscountInfo?: string;
    embeddedNoteInfo?: string;
    embeddedShipmentProviderInfo?: string;
    embeddedVipQuotaInfo?: string;
    etaInfo?: EtaInfo;
    externalTransactionId?: string;
    first?: boolean;
    firstName?: string;
    groupId?: number;
    id?: number;
    invoiceAddressDetails?: string;
    invoiceAddressId?: number;
    invoiceNumber?: string;
    invoicedPrice?: number;
    ip?: string;
    itemCount?: number;
    lastName?: string;
    lastPaymentMessage?: string;
    lastTimeSlotOfDay?: boolean;
    lastUpdateOperation?: OrderDTO.LastUpdateOperationEnum;
    lastUpdatedAt?: string;
    lastUpdaterId?: number;
    latestCollectingStatus?: string;
    loyaltyCard?: string;
    mainPaymentOption?: PaymentOption;
    mainPaymentType?: OrderDTO.MainPaymentTypeEnum;
    maximumAmount?: number;
    minimumRequiredTotalPrice?: number;
    note?: string;
    orderRate?: number;
    origin?: string;
    parentId?: number;
    paymentStoreId?: number;
    phoneNumber?: string;
    polygonId?: number;
    refundedPrice?: number;
    regionId?: number;
    revenue?: number;
    scheduleType?: OrderDTO.ScheduleTypeEnum;
    serviceAreaName?: string;
    serviceAreaObjectId?: number;
    serviceAreaObjectType?: string;
    shadow?: boolean;
    shipmentProvider?: string;
    shoppingType?: string;
    simpleOrderFeedback?: SimpleOrderFeedbackDTO;
    staff?: boolean;
    status?: OrderDTO.StatusEnum;
    storeId?: number;
    storePortfolio?: OrderDTO.StorePortfolioEnum;
    timeSlotDetails?: string;
    timeSlotId?: number;
    transferredStoreId?: number;
    type?: OrderDTO.TypeEnum;
    userId?: number;
    vip?: boolean;
}
export namespace OrderDTO {
    export type AlternativeProductChoiceEnum = 'NO_ALTERNATIVE' | 'UP_TO_CUSTOMER' | 'UP_TO_PERSONNEL';
    export const AlternativeProductChoiceEnum = {
        NoAlternative: 'NO_ALTERNATIVE' as AlternativeProductChoiceEnum,
        UpToCustomer: 'UP_TO_CUSTOMER' as AlternativeProductChoiceEnum,
        UpToPersonnel: 'UP_TO_PERSONNEL' as AlternativeProductChoiceEnum
    };
    export type ChannelEnum = 'FUUDY' | 'GLOVO' | 'INTERNAL' | 'TRENDYOL';
    export const ChannelEnum = {
        Fuudy: 'FUUDY' as ChannelEnum,
        Glovo: 'GLOVO' as ChannelEnum,
        Internal: 'INTERNAL' as ChannelEnum,
        Trendyol: 'TRENDYOL' as ChannelEnum
    };
    export type CollectionStatusEnum = 'ASSIGNED_FOR_DELIVERY' | 'CANCELLED' | 'COLLECTED' | 'COLLECTING' | 'COLLECTING_PAUSED' | 'COLLECTION_APPROVED' | 'DELIVERED' | 'INVOICED' | 'IN_DELIVERY' | 'IN_POOL' | 'NEW_PENDING' | 'READY_FOR_DELIVERY';
    export const CollectionStatusEnum = {
        AssignedForDelivery: 'ASSIGNED_FOR_DELIVERY' as CollectionStatusEnum,
        Cancelled: 'CANCELLED' as CollectionStatusEnum,
        Collected: 'COLLECTED' as CollectionStatusEnum,
        Collecting: 'COLLECTING' as CollectionStatusEnum,
        CollectingPaused: 'COLLECTING_PAUSED' as CollectionStatusEnum,
        CollectionApproved: 'COLLECTION_APPROVED' as CollectionStatusEnum,
        Delivered: 'DELIVERED' as CollectionStatusEnum,
        Invoiced: 'INVOICED' as CollectionStatusEnum,
        InDelivery: 'IN_DELIVERY' as CollectionStatusEnum,
        InPool: 'IN_POOL' as CollectionStatusEnum,
        NewPending: 'NEW_PENDING' as CollectionStatusEnum,
        ReadyForDelivery: 'READY_FOR_DELIVERY' as CollectionStatusEnum
    };
    export type DeliveryModelEnum = 'LAST_MILE' | 'SHIPMENT';
    export const DeliveryModelEnum = {
        LastMile: 'LAST_MILE' as DeliveryModelEnum,
        Shipment: 'SHIPMENT' as DeliveryModelEnum
    };
    export type LastUpdateOperationEnum = 'ADMIN_UPDATE' | 'CHANGE_STORE' | 'RESCHEDULE' | 'UPDATE' | 'UPDATE_INVOICE_ADDRESS';
    export const LastUpdateOperationEnum = {
        AdminUpdate: 'ADMIN_UPDATE' as LastUpdateOperationEnum,
        ChangeStore: 'CHANGE_STORE' as LastUpdateOperationEnum,
        Reschedule: 'RESCHEDULE' as LastUpdateOperationEnum,
        Update: 'UPDATE' as LastUpdateOperationEnum,
        UpdateInvoiceAddress: 'UPDATE_INVOICE_ADDRESS' as LastUpdateOperationEnum
    };
    export type MainPaymentTypeEnum = 'BKM' | 'CARD_REWARD' | 'CASH_ON_DELIVERY' | 'COUPON' | 'CREDIT_CARD' | 'CREDIT_CARD_ON_DELIVERY' | 'CUSTOMER_BOND' | 'GARANTI_PAY' | 'INSTANT_DISCOUNT' | 'LOAN' | 'MASTERPASS' | 'MONEY_POINT' | 'PERSONNEL_BOND' | 'VALUE_DATE' | 'VOUCHER' | 'WALLET';
    export const MainPaymentTypeEnum = {
        Bkm: 'BKM' as MainPaymentTypeEnum,
        CardReward: 'CARD_REWARD' as MainPaymentTypeEnum,
        CashOnDelivery: 'CASH_ON_DELIVERY' as MainPaymentTypeEnum,
        Coupon: 'COUPON' as MainPaymentTypeEnum,
        CreditCard: 'CREDIT_CARD' as MainPaymentTypeEnum,
        CreditCardOnDelivery: 'CREDIT_CARD_ON_DELIVERY' as MainPaymentTypeEnum,
        CustomerBond: 'CUSTOMER_BOND' as MainPaymentTypeEnum,
        GarantiPay: 'GARANTI_PAY' as MainPaymentTypeEnum,
        InstantDiscount: 'INSTANT_DISCOUNT' as MainPaymentTypeEnum,
        Loan: 'LOAN' as MainPaymentTypeEnum,
        Masterpass: 'MASTERPASS' as MainPaymentTypeEnum,
        MoneyPoint: 'MONEY_POINT' as MainPaymentTypeEnum,
        PersonnelBond: 'PERSONNEL_BOND' as MainPaymentTypeEnum,
        ValueDate: 'VALUE_DATE' as MainPaymentTypeEnum,
        Voucher: 'VOUCHER' as MainPaymentTypeEnum,
        Wallet: 'WALLET' as MainPaymentTypeEnum
    };
    export type ScheduleTypeEnum = 'INSTANT' | 'SHIPMENT' | 'TIME_SLOT';
    export const ScheduleTypeEnum = {
        Instant: 'INSTANT' as ScheduleTypeEnum,
        Shipment: 'SHIPMENT' as ScheduleTypeEnum,
        TimeSlot: 'TIME_SLOT' as ScheduleTypeEnum
    };
    export type StatusEnum = 'AUTH_ERROR' | 'AUTH_READY' | 'COMPLETED' | 'CREATED' | 'IN_AUTH' | 'IN_DELIVERY';
    export const StatusEnum = {
        AuthError: 'AUTH_ERROR' as StatusEnum,
        AuthReady: 'AUTH_READY' as StatusEnum,
        Completed: 'COMPLETED' as StatusEnum,
        Created: 'CREATED' as StatusEnum,
        InAuth: 'IN_AUTH' as StatusEnum,
        InDelivery: 'IN_DELIVERY' as StatusEnum
    };
    export type StorePortfolioEnum = 'MARKET' | 'ELECTRONIC' | 'KURBAN' | 'CUSTOM';
    export const StorePortfolioEnum = {
        Market: 'MARKET' as StorePortfolioEnum,
        Electronic: 'ELECTRONIC' as StorePortfolioEnum,
        Kurban: 'KURBAN' as StorePortfolioEnum,
        Custom: 'CUSTOM' as StorePortfolioEnum
    };
    export type TypeEnum = 'ADDITIONAL' | 'MAIN' | 'UPDATE';
    export const TypeEnum = {
        Additional: 'ADDITIONAL' as TypeEnum,
        Main: 'MAIN' as TypeEnum,
        Update: 'UPDATE' as TypeEnum
    };
}
