// @migroscomtr/model/simpleStoreProductInfoDTO.model.ts
/**
 * Api Documentation
 * Api Documentation
 *
 * The version of the OpenAPI document: 1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { StoreProductBadge } from './storeProductBadge.model';
import { TagInfo1 } from './tagInfo1.model';
import { ImageUrls } from './imageUrls.model';
import { ProductBrandInfo } from './productBrandInfo.model';
import { ProductCategoryInfo } from './productCategoryInfo.model';


export interface SimpleStoreProductInfoDTO { 
    alternativeUnit?: SimpleStoreProductInfoDTO.AlternativeUnitEnum;
    alternativeUnitIncrementAmount?: number;
    alternativeUnitInitialIncrementAmount?: number;
    alternativeUnitMaxAmount?: number;
    alternativeUnitValue?: number;
    badges?: Array<StoreProductBadge>;
    brand?: ProductBrandInfo;
    brandName?: string;
    categoryAscendants?: Array<ProductCategoryInfo>;
    crmDiscountTags?: Array<TagInfo1>;
    id?: number;
    images?: Array<ImageUrls>;
    incrementAmount?: number;
    initialIncrementAmount?: number;
    maxAmount?: number;
    name?: string;
    prettyName?: string;
    primaryUnit?: SimpleStoreProductInfoDTO.PrimaryUnitEnum;
    referrerEventId?: string;
    regularPrice?: number;
    shownPrice?: number;
    sku?: string;
    status?: SimpleStoreProductInfoDTO.StatusEnum;
    storeId?: number;
    storePortfolio?: SimpleStoreProductInfoDTO.StorePortfolioEnum;
    unit?: SimpleStoreProductInfoDTO.UnitEnum;
    useOnlyAlternativeUnit?: boolean;
    warehouseId?: number;
}
export namespace SimpleStoreProductInfoDTO {
    export type AlternativeUnitEnum = 'BUNCH' | 'CUBIC_MILLIMETER' | 'GRAM' | 'MILLILITER' | 'MILLIMETER' | 'PACKET' | 'PIECE' | 'ROLL' | 'SACK' | 'SET' | 'SQUARE_MILLIMETER' | 'TOME';
    export const AlternativeUnitEnum = {
        Bunch: 'BUNCH' as AlternativeUnitEnum,
        CubicMillimeter: 'CUBIC_MILLIMETER' as AlternativeUnitEnum,
        Gram: 'GRAM' as AlternativeUnitEnum,
        Milliliter: 'MILLILITER' as AlternativeUnitEnum,
        Millimeter: 'MILLIMETER' as AlternativeUnitEnum,
        Packet: 'PACKET' as AlternativeUnitEnum,
        Piece: 'PIECE' as AlternativeUnitEnum,
        Roll: 'ROLL' as AlternativeUnitEnum,
        Sack: 'SACK' as AlternativeUnitEnum,
        Set: 'SET' as AlternativeUnitEnum,
        SquareMillimeter: 'SQUARE_MILLIMETER' as AlternativeUnitEnum,
        Tome: 'TOME' as AlternativeUnitEnum
    };
    export type PrimaryUnitEnum = 'BUNCH' | 'CUBIC_MILLIMETER' | 'GRAM' | 'MILLILITER' | 'MILLIMETER' | 'PACKET' | 'PIECE' | 'ROLL' | 'SACK' | 'SET' | 'SQUARE_MILLIMETER' | 'TOME';
    export const PrimaryUnitEnum = {
        Bunch: 'BUNCH' as PrimaryUnitEnum,
        CubicMillimeter: 'CUBIC_MILLIMETER' as PrimaryUnitEnum,
        Gram: 'GRAM' as PrimaryUnitEnum,
        Milliliter: 'MILLILITER' as PrimaryUnitEnum,
        Millimeter: 'MILLIMETER' as PrimaryUnitEnum,
        Packet: 'PACKET' as PrimaryUnitEnum,
        Piece: 'PIECE' as PrimaryUnitEnum,
        Roll: 'ROLL' as PrimaryUnitEnum,
        Sack: 'SACK' as PrimaryUnitEnum,
        Set: 'SET' as PrimaryUnitEnum,
        SquareMillimeter: 'SQUARE_MILLIMETER' as PrimaryUnitEnum,
        Tome: 'TOME' as PrimaryUnitEnum
    };
    export type StatusEnum = 'DELIST' | 'IN_SALE' | 'STOCK_OUT';
    export const StatusEnum = {
        Delist: 'DELIST' as StatusEnum,
        InSale: 'IN_SALE' as StatusEnum,
        StockOut: 'STOCK_OUT' as StatusEnum
    };
    export type StorePortfolioEnum = 'MARKET' | 'ELECTRONIC' | 'KURBAN' | 'CUSTOM';
    export const StorePortfolioEnum = {
        Market: 'MARKET' as StorePortfolioEnum,
        Electronic: 'ELECTRONIC' as StorePortfolioEnum,
        Kurban: 'KURBAN' as StorePortfolioEnum,
        Custom: 'CUSTOM' as StorePortfolioEnum
    };
    export type UnitEnum = 'BUNCH' | 'CUBIC_MILLIMETER' | 'GRAM' | 'MILLILITER' | 'MILLIMETER' | 'PACKET' | 'PIECE' | 'ROLL' | 'SACK' | 'SET' | 'SQUARE_MILLIMETER' | 'TOME';
    export const UnitEnum = {
        Bunch: 'BUNCH' as UnitEnum,
        CubicMillimeter: 'CUBIC_MILLIMETER' as UnitEnum,
        Gram: 'GRAM' as UnitEnum,
        Milliliter: 'MILLILITER' as UnitEnum,
        Millimeter: 'MILLIMETER' as UnitEnum,
        Packet: 'PACKET' as UnitEnum,
        Piece: 'PIECE' as UnitEnum,
        Roll: 'ROLL' as UnitEnum,
        Sack: 'SACK' as UnitEnum,
        Set: 'SET' as UnitEnum,
        SquareMillimeter: 'SQUARE_MILLIMETER' as UnitEnum,
        Tome: 'TOME' as UnitEnum
    };
}
