// @migroscomtr/model/storeProductInfoDTO.model.ts
/**
 * Api Documentation
 * Api Documentation
 *
 * The version of the OpenAPI document: 1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { ProductPropertyInfo } from './productPropertyInfo.model';
import { StoreProductBadge } from './storeProductBadge.model';
import { TagInfo1 } from './tagInfo1.model';
import { ImageUrls } from './imageUrls.model';
import { ProductListInfo } from './productListInfo.model';
import { ProductBrandInfo } from './productBrandInfo.model';
import { ProductCategoryInfo } from './productCategoryInfo.model';


export interface StoreProductInfoDTO { 
    alternativeUnit?: StoreProductInfoDTO.AlternativeUnitEnum;
    alternativeUnitIncrementAmount?: number;
    alternativeUnitInitialIncrementAmount?: number;
    alternativeUnitMaxAmount?: number;
    alternativeUnitValue?: number;
    badges?: Array<StoreProductBadge>;
    brand?: ProductBrandInfo;
    brandId?: number;
    category?: ProductCategoryInfo;
    categoryAscendants?: Array<ProductCategoryInfo>;
    categoryId?: number;
    crmDiscountTags?: Array<TagInfo1>;
    description?: string;
    discountDescriptions?: Array<string>;
    discounts?: { [key: string]: Array<string>; };
    id?: number;
    images?: Array<ImageUrls>;
    incrementAmount?: number;
    initialIncrementAmount?: number;
    legalDescription?: string;
    lists?: Array<ProductListInfo>;
    loyaltyPrice?: number;
    maxAmount?: number;
    name?: string;
    prettyName?: string;
    properties?: { [key: string]: { [key: string]: string; }; };
    propertyInfosMap?: { [key: string]: Array<ProductPropertyInfo>; };
    referrerEventId?: string;
    regularPrice?: number;
    salePrice?: number;
    saleable?: boolean;
    seoDescription?: string;
    shownPrice?: number;
    sku?: string;
    sortedProperties?: { [key: string]: Array<ProductPropertyInfo>; };
    specialDietTags?: Array<TagInfo1>;
    status?: StoreProductInfoDTO.StatusEnum;
    storeId?: number;
    storePortfolio?: StoreProductInfoDTO.StorePortfolioEnum;
    unit?: StoreProductInfoDTO.UnitEnum;
    unitAmount?: number;
    useOnlyAlternativeUnit?: boolean;
    warehouseId?: number;
}
export namespace StoreProductInfoDTO {
    export type AlternativeUnitEnum = 'BUNCH' | 'CUBIC_MILLIMETER' | 'GRAM' | 'MILLILITER' | 'MILLIMETER' | 'PACKET' | 'PIECE' | 'ROLL' | 'SACK' | 'SET' | 'SQUARE_MILLIMETER' | 'TOME';
    export const AlternativeUnitEnum = {
        Bunch: 'BUNCH' as AlternativeUnitEnum,
        CubicMillimeter: 'CUBIC_MILLIMETER' as AlternativeUnitEnum,
        Gram: 'GRAM' as AlternativeUnitEnum,
        Milliliter: 'MILLILITER' as AlternativeUnitEnum,
        Millimeter: 'MILLIMETER' as AlternativeUnitEnum,
        Packet: 'PACKET' as AlternativeUnitEnum,
        Piece: 'PIECE' as AlternativeUnitEnum,
        Roll: 'ROLL' as AlternativeUnitEnum,
        Sack: 'SACK' as AlternativeUnitEnum,
        Set: 'SET' as AlternativeUnitEnum,
        SquareMillimeter: 'SQUARE_MILLIMETER' as AlternativeUnitEnum,
        Tome: 'TOME' as AlternativeUnitEnum
    };
    export type StatusEnum = 'DELIST' | 'IN_SALE' | 'STOCK_OUT';
    export const StatusEnum = {
        Delist: 'DELIST' as StatusEnum,
        InSale: 'IN_SALE' as StatusEnum,
        StockOut: 'STOCK_OUT' as StatusEnum
    };
    export type StorePortfolioEnum = 'MARKET' | 'ELECTRONIC' | 'KURBAN' | 'CUSTOM';
    export const StorePortfolioEnum = {
        Market: 'MARKET' as StorePortfolioEnum,
        Electronic: 'ELECTRONIC' as StorePortfolioEnum,
        Kurban: 'KURBAN' as StorePortfolioEnum,
        Custom: 'CUSTOM' as StorePortfolioEnum
    };
    export type UnitEnum = 'BUNCH' | 'CUBIC_MILLIMETER' | 'GRAM' | 'MILLILITER' | 'MILLIMETER' | 'PACKET' | 'PIECE' | 'ROLL' | 'SACK' | 'SET' | 'SQUARE_MILLIMETER' | 'TOME';
    export const UnitEnum = {
        Bunch: 'BUNCH' as UnitEnum,
        CubicMillimeter: 'CUBIC_MILLIMETER' as UnitEnum,
        Gram: 'GRAM' as UnitEnum,
        Milliliter: 'MILLILITER' as UnitEnum,
        Millimeter: 'MILLIMETER' as UnitEnum,
        Packet: 'PACKET' as UnitEnum,
        Piece: 'PIECE' as UnitEnum,
        Roll: 'ROLL' as UnitEnum,
        Sack: 'SACK' as UnitEnum,
        Set: 'SET' as UnitEnum,
        SquareMillimeter: 'SQUARE_MILLIMETER' as UnitEnum,
        Tome: 'TOME' as UnitEnum
    };
}
