// @migroscomtr/model/timeSlotInfoDto.model.ts
/**
 * Api Documentation
 * Api Documentation
 *
 * The version of the OpenAPI document: 1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export interface TimeSlotInfoDto { 
    active?: boolean;
    day?: TimeSlotInfoDto.DayEnum;
    deliveryPrice?: number;
    disabledPayments?: string;
    finishHour?: number;
    id?: number;
    lastOrderMinute?: number;
    regionId?: number;
    regionTimeSlotId?: number;
    regularDeliveryPrice?: number;
    startHour?: number;
    status?: TimeSlotInfoDto.StatusEnum;
    timeSlot?: string;
    vipQuota?: boolean;
}
export namespace TimeSlotInfoDto {
    export type DayEnum = 'FRIDAY' | 'MONDAY' | 'SATURDAY' | 'SUNDAY' | 'THURSDAY' | 'TUESDAY' | 'WEDNESDAY';
    export const DayEnum = {
        Friday: 'FRIDAY' as DayEnum,
        Monday: 'MONDAY' as DayEnum,
        Saturday: 'SATURDAY' as DayEnum,
        Sunday: 'SUNDAY' as DayEnum,
        Thursday: 'THURSDAY' as DayEnum,
        Tuesday: 'TUESDAY' as DayEnum,
        Wednesday: 'WEDNESDAY' as DayEnum
    };
    export type StatusEnum = 'AVAILABLE' | 'CAPACITY_EMPTY' | 'CAPACITY_FILLED' | 'CLOSED' | 'NOT_IN_SERVICE_HOURS' | 'START_HOUR_PASSED';
    export const StatusEnum = {
        Available: 'AVAILABLE' as StatusEnum,
        CapacityEmpty: 'CAPACITY_EMPTY' as StatusEnum,
        CapacityFilled: 'CAPACITY_FILLED' as StatusEnum,
        Closed: 'CLOSED' as StatusEnum,
        NotInServiceHours: 'NOT_IN_SERVICE_HOURS' as StatusEnum,
        StartHourPassed: 'START_HOUR_PASSED' as StatusEnum
    };
}
