// @migroscomtr/model/workaroundBalanceDTO.model.ts
/**
 * Api Documentation
 * Api Documentation
 *
 * The version of the OpenAPI document: 1.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export interface WorkaroundBalanceDTO { 
    balanceCode?: WorkaroundBalanceDTO.BalanceCodeEnum;
    code?: WorkaroundBalanceDTO.CodeEnum;
    minValue?: number;
    unitCode?: WorkaroundBalanceDTO.UnitCodeEnum;
    value?: number;
}
export namespace WorkaroundBalanceDTO {
    export type BalanceCodeEnum = 'MONEY' | 'PERSONELBONO' | 'MPRO';
    export const BalanceCodeEnum = {
        Money: 'MONEY' as BalanceCodeEnum,
        Personelbono: 'PERSONELBONO' as BalanceCodeEnum,
        Mpro: 'MPRO' as BalanceCodeEnum
    };
    export type CodeEnum = 'MONEY' | 'PERSONELBONO' | 'MPRO';
    export const CodeEnum = {
        Money: 'MONEY' as CodeEnum,
        Personelbono: 'PERSONELBONO' as CodeEnum,
        Mpro: 'MPRO' as CodeEnum
    };
    export type UnitCodeEnum = 'BEBEMONEY' | 'MILE' | 'MONEY' | 'PARACIK' | 'TL';
    export const UnitCodeEnum = {
        Bebemoney: 'BEBEMONEY' as UnitCodeEnum,
        Mile: 'MILE' as UnitCodeEnum,
        Money: 'MONEY' as UnitCodeEnum,
        Paracik: 'PARACIK' as UnitCodeEnum,
        Tl: 'TL' as UnitCodeEnum
    };
}
